src/cls/rgw/cls_rgw.cc

```diff
@@ -640,19 +631,41 @@ int rgw_bucket_list(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
 		__func__, escape_str(kiter->first).c_str());
         continue;
       }
+      CLS_LOG(20, "%s: working on key=%s len=%zu",
+	      __func__, kiter->first.c_str(), kiter->first.size());
+      start_after_omap_key = kiter->first;
 
-      if (!entry.is_valid()) {
-        CLS_LOG(20, "%s: entry %s[%s] is not valid",
-		__func__, key.name.c_str(), key.instance.c_str());
-        continue;
-      }
+      // these checks are performed also by RGW
+      if (rgw_do_osd_side_checks) {
```

Comment:
The cost of these checks could be lowered if the flags field in the encoding schema move at the beginning. Yet, this would require using a feature (/ release) bit.